addon-dev: incremental updates to output

patricklx opened the pull request:
this now only updates changed files in dist. And deletes removed files in dist.
otherwise all files are deleted on each change, which makes HMR less useful

NullVoxPopuli reviewed (dismissed):
Logically, it seems good.
Do we have any tests for this behavior? I don't remember: 😅

mansona reviewed (changes requested):
so one thing that I would like to do before we actually move forward on this is that we verify that this Plugin works correctly with the bug scenario that https://github.com/embroider-build/embroider/pull/1448 was originally trying to fix. Ideally we would cover this in the v2 addon dev watch tests

patricklx commented:
> so one thing that I would like to do before we actually move forward on this is that we verify that this Plugin works correctly with the bug scenario that #1448 was originally trying to fix. Ideally we would cover this in the v2 addon dev watch tests

I think what this actually does now is more like a `sync`. When a file gets deleted in src, the corresponding generated file in dist also gets deleted.
there is the option `runOnce`, which only deletes the files only once at start. Which is already better, but it leaves the deleted files during dev.

mansona commented:
> I think this also fixes the original issue

The problem I have with this PR is that I want more confidence that it fixes the original issue before we move forward with it. Myself and @NullVoxPopuli tried to recreate the issue on the Triage meeting and we couldn't get it to fail so we can't tell if this makes any difference 🙃 

I would really like a failing test that shows the problem in a repeatable way before we try to fix it by altering the underlying implementation. If this is not possible (because maybe it's an OS level race condition 🤷) then at the very least we need to coordinate with @simonihmig because he said he would be looking into the original issue over the next while: https://github.com/embroider-build/addon-blueprint/issues/32#issuecomment-2079582665

patricklx commented:
I'm not sure how to add a test for this

patricklx commented:
@mansona i improved this even more and added some tests

simonihmig commented on packages/addon-dev/src/rollup-incremental-plugin.ts:
```diff
@@ -0,0 +1,70 @@
+        if (
+          bundle[checkKey]?.type === 'asset' &&
+          (bundle[checkKey] as OutputAsset).source ===
+            generatedAssets.get(checkKey)
```
Trying this locally, this if clause was not catching a lot of asset files as expected, causing them to be written to disk even when unchanged. The issue seems to be that depending on the file type, `.source` could point to a Buffer, in which case you would have to compare them via [.equals()](https://nodejs.org/api/buffer.html#bufequalsotherbuffer).

simonihmig commented on packages/addon-dev/src/rollup-incremental-plugin.ts:
```diff
@@ -0,0 +1,70 @@
+          delete bundle[key];
+          continue;
+        } else {
+          generatedAssets.set(
```
This else block would also be executed for non-asset files, which is not expected, right?
I would think there should be an out-if checking of the current file is an asset, and another inner-if that checks for quality, so that this block is only reachable for assets that are not equal to the previous run.

simonihmig reviewed (changes requested):
I tested this locally with a real project and `yarn link`. I found some minor issues, see inline comments. But the good part is that this seems to also fix some of the issues discussed in https://github.com/embroider-build/addon-blueprint/issues/32#issuecomment-2079582665.

The main issue I was hoping to get fixed by a revised clean plugin was the fact that an app watching the dist folder of an addon would often pick up changes as they were written but before the addon's rebuild has finished. 
Leading to multiple app rebuilds, with the first ones often picking up interim build errors. Which would quickly go away, when the addon's rebuild has finished, but that's still annoying...

And testing this change here locally indeed seems to fix that issue! :tada:

simonihmig commented on packages/addon-dev/src/rollup.ts:
```diff
@@ -64,10 +63,11 @@ export class Addon {
+  // rollup already supports incremental transforms of files,
+  // this extends it to the dist files
+  clean() {
+    return clean();
```
If we consider this breaking, we could still release this on stable as a minor version, by exposing this as a separate plugin, and updating the blueprint to use the new one instead of `addon.clean()`! 🤔

patricklx commented on tests/scenarios/v2-addon-dev-watch-test.ts:
```diff
@@ -49,15 +50,15 @@ Scenarios.fromProject(() => baseV2Addon())
-            addon.publicEntrypoints(['components/**/*.js']),
-            addon.appReexports(['components/**/*.js']),
+            addon.publicEntrypoints(['components/**/*.{gts,js}']),
+            addon.appReexports(['components/**/*.{gts,js}']),
```
Ah, maybe it wasn't needed

patricklx commented on packages/addon-dev/src/rollup-incremental-plugin.ts:
```diff
@@ -0,0 +1,70 @@
+    },
+    generateBundle(options, bundle) {
+      if (existsSync(options.dir!)) {
+        const files = walkSync(options.dir!, {
```
It's required to delete files that are not part of the bundle

patricklx replied:
I improved this ti only walk once

patricklx commented on packages/addon-dev/src/rollup-incremental-plugin.ts:
```diff
@@ -0,0 +1,70 @@
+        for (const file of files) {
+          if (!bundle[file]) {
+            generatedAssets.delete(file);
+            rmSync(join(options.dir!, file));
```
generated assets do not pass through transform. So it's not detectable

patricklx commented on packages/addon-dev/src/rollup-gjs-plugin.ts:
```diff
@@ -14,11 +13,10 @@ export default function rollupGjsPlugin(
     name: PLUGIN_NAME,
 
-    load(id: string) {
+    transform(input: string, id: string) {
```
It's required to have hbs pass through transform to have thrm incremental.

ef4 replied:
I'm glad to use `transform` over `load` whenever possible. `transform` composes more easily with other plugins. 👍

simonihmig commented on packages/addon-dev/src/rollup-incremental-plugin.ts:
```diff
@@ -42,14 +53,22 @@ export default function incremental(): Plugin {
             generatedAssets.set(
               checkKey,
-              (bundle[checkKey] as OutputAsset).source
+              (bundle[checkKey] as OutputAsset).source.toString()
```
Isn't that string casting breaking your `isEqual` implementation? Like when the plugin sees a jpeg asset, `(bundle[checkKey] as OutputAsset).source` will be a Buffer, but you put it into `generatedAssets` as a string, so `isEqual` will always return false, won't it?

patricklx replied:
your right! thanks

patricklx commented on tests/scenarios/scenarios.ts:
```diff
@@ -1,5 +1,28 @@
 import { dirname } from 'path';
+import { exec } from 'child_process';
+
+let prepare = Scenario.prototype.prepare;
```
waiting on https://github.com/stefanpenner/node-fixturify-project/pull/98

ef4 commented on packages/addon-dev/src/rollup-hbs-plugin.ts:
```diff
@@ -30,30 +29,34 @@ export default function rollupHbsPlugin({
+  import { setComponentTemplate } from '@ember/component';
+  export default setComponentTemplate(precompileTemplate(\`${code}\`), templateOnly());
+  `;
+}
```
We were avoiding this kind of inlining on purpose. I would prefer to keep letting rollup do the inlining automatically.

patricklx replied:
I do not see how it can be done when using `transform` in the hbs plugin.

the hbs gets transformed by embroiders babel plugin to
e.g. if we get the file `demo.hbs`
we would have
```js
import { setComponentTemplate } from '@ember/component';
import default as TEMPLATE from './demo.hbs'
export default setComponentTemplate(precompileTemplate(TEMPLATE), templateOnly());
```

but ` './demo.hbs'` is already cached with the transform, so we would get the same thing back

patricklx replied:
i fixed it

ef4 commented on packages/addon-dev/src/rollup-incremental-plugin.ts:
```diff
@@ -0,0 +1,111 @@
+      // could also be done directly in the babel plugin
+      // by passing rollup context into it
+      let hbsFilename = id.replace(/\.\w{1,3}$/, '') + '.hbs';
+      if (hbsFilename !== id && existsSync(hbsFilename)) {
```
This feature doesn't belong here, because the dependency was created by the hbs plugin but is being patched in by the incremental plugin.

And the only reason you need this is the manual inlining that I mentioned above. It should not be needed at all.

patricklx replied:
i moved it to the hbs plugin. adding watch file is required when deleting or creating files that take part in the colocation

simonihmig commented on tests/scenarios/compat-stage2-test.ts:
```diff
@@ -40,6 +40,9 @@ stage2Scenarios
       },
     });
+
+    (depC.pkg['ember-addon'] as any).before = ['dep-a', 'dep-b', 'in-repo-d', 'in-repo-a'];
```
Why do we have this?

patricklx replied:
Using windows real path instead of short path alias caused this

simonihmig commented on tests/scenarios/v2-addon-dev-test.ts:
```diff
@@ -73,6 +73,8 @@ appScenarios
               exclude: ['**/-excluded/**/*'],
             }),
 
+            addon.clean(),
```
Is this change in order required? If so, then this is clearly a breaking change!

patricklx replied:
I think it is. I need to revalidate

simonihmig replied:
Any news on this @patricklx? 

With ~250 addons in our monorepo (not all v2 though), I'd rather not have to write a codemod to change the order! 😏

patricklx replied:
hi @simonihmig I reverted the change and the tests passed. (other unrelated failed)

patricklx replied:
hold that, i forgot one, lets see

simonihmig replied:
There was a timeout in one test job. I restarted and it is all green now! :tada: 

Thanks for getting back to this, I think this revert is helpful, as it shows we are not introducing breaking changes here!

simonihmig commented on tests/scenarios/v2-addon-dev-watch-test.ts:
```diff
@@ -134,6 +137,139 @@ Scenarios.fromProject(() => baseV2Addon())
+            },
+          });
+
+          // updating hbs content should update resulting app re-exported component
```
Is this a typo, or am I misunderstanding?

```suggestion
          // updating hbs content should not update resulting app re-exported component
```

simonihmig commented on tests/scenarios/v2-addon-dev-watch-test.ts:
```diff
@@ -134,6 +137,139 @@ Scenarios.fromProject(() => baseV2Addon())
+          await fs.writeFile(demoJs, demoJsContent);
+          await watcher?.nextBuild();
+          distPathDemoCompContent = await fs.readFile(distPathDemoComp);
+          assert.ok(!distPathDemoCompContent.includes('templateOnly'));
```
nit pick:

```suggestion
          assert.false(distPathDemoCompContent.includes('templateOnly'));
```

(this was initially confusing to me, as it took me a while to see the `!` 😅 )

simonihmig commented on tests/scenarios/v2-addon-dev-watch-test.ts:
```diff
@@ -134,6 +137,139 @@ Scenarios.fromProject(() => baseV2Addon())
+          await fs.rm(demoJs);
+          await watcher?.nextBuild();
+          let distPathDemoCompContent = await fs.readFile(distPathDemoComp);
+          assert.ok(distPathDemoCompContent.includes('templateOnly'));
```
nit pick:

```suggestion
          assert.true(distPathDemoCompContent.includes('templateOnly'));
```

simonihmig commented on tests/scenarios/v2-addon-dev-watch-test.ts:
```diff
@@ -134,6 +137,139 @@ Scenarios.fromProject(() => baseV2Addon())
+
+          // updating template only hbs should update the dist output
+          distPath = path.join(addon.dir, 'dist/components/button.js');
+          await isNotModified({
```
why is this _not_ getting modified?

patricklx replied:
that's actually correct. The rexport file did not change.
Wrong comment

patricklx replied:
it should update---

ef4 commented on packages/addon-dev/src/rollup.ts:
```diff
@@ -2,14 +2,13 @@ import { default as hbs } from './rollup-hbs-plugin';
 import { default as publicEntrypoints } from './rollup-public-entrypoints';
 import { default as appReexports } from './rollup-app-reexports';
-import type { Options as DelOptions } from 'rollup-plugin-delete';
-import { default as clean } from 'rollup-plugin-delete';
```
Let's remove this from package.json. I think this was probably the only use.

ef4 commented on packages/addon-dev/src/rollup-hbs-plugin.ts:
```diff
@@ -30,9 +30,24 @@ export default function rollupHbsPlugin({
+      if (hbsFilter(id)) {
+        let basename = id.replace(/\.\w{1,3}$/, '');
+        this.addWatchFile(basename + '.ts');
+        this.addWatchFile(basename + '.js');
```
I think these belong beside the spot where we actually attempted to resolve the corresponding JS files in `maybeSynthesizeJS`. Having them together may address the caching behaviors you mentioned that caused you to have to move code between load and transform phases.

patricklx replied:
I tried to move it to the resolve hook. But then the transform for js files does not get updated. It might need to be in the transform hook for the cache to be invalidated

simonihmig reviewed (approved):
LGTM!

I think we are close to getting this merged finally,. Thanks for working on this, Patrick, I am looking forward to get to use this!

@ef4 @mansona @NullVoxPopuli wanna give this a final review?

AskerFED commented:
![image](https://github.com/user-attachments/assets/a0084443-c524-453a-8ff3-141eceb79e66)

I have faced the ENOENT error for the addon.clean() which triggers file deletion event while rollup watch but it is not build stopping issue just error thrown and started build once dist was replaced. After @NullVoxPopuli suggestion to try this addon-dev branch now ENOENT error resolved. But the below warning is shown but is harmless I guess...

![image](https://github.com/user-attachments/assets/9a87e36d-77d3-4658-86b2-20af579132c1)

patricklx commented on packages/addon-dev/src/rollup-gjs-plugin.ts:
```diff
@@ -14,11 +13,10 @@ export default function rollupGjsPlugin(
     name: PLUGIN_NAME,
 
-    load(id: string) {
+    transform(input: string, id: string) {
```
@ef4 I noticed that now the gjs plugin must come before the babel plugin in the rollup config. That would make this a breaking change

patricklx commented on packages/addon-dev/src/rollup-hbs-plugin.ts:
```diff
@@ -12,6 +12,10 @@ export default function rollupHbsPlugin({
     async resolveId(source: string, importer: string | undefined, options) {
+      let hbsFilename = source.replace(/\.\w{1,3}$/, '') + '.hbs';
+      if (hbsFilename !== source) {
+        this.addWatchFile(hbsFilename);
```
@ef4 I had to add addWatchFile to all 3 methods to get this to work correctly on windows

ef4 commented:
This is hard for me to follow. Can we take advantage of the cleanup that has happened in the vite hbs plugin to use the same patterns here?

patricklx commented:
> This is hard for me to follow. Can we take advantage of the cleanup that has happened in the vite hbs plugin to use the same patterns here?

I updated the hbs plugin to be similar to the one in vite hbs.
only a with a few changes. there is no generic extension search setup, it could be setup.
I added the extension search for .hbs extension.

also added the addWatchFile calls to get tests working

patricklx commented on packages/addon-dev/src/rollup-hbs-plugin.ts:
```diff
@@ -12,48 +19,88 @@ export default function rollupHbsPlugin({
+        }
+      }
+
+      let syntheticId = needsSyntheticComponentJS(source, resolution.id);
```
we do not have a packageCache here

patricklx commented on packages/addon-dev/src/rollup-hbs-plugin.ts:
```diff
@@ -63,6 +63,7 @@ export default function rollupHbsPlugin({
       let syntheticId = needsSyntheticComponentJS(source, resolution.id);
       if (syntheticId) {
         this.addWatchFile(source);
+        this.addWatchFile(syntheticId);
```
this change did not work @ef4 .
I think transform does have its own specific cache.
https://rollupjs.org/plugin-development/#transform

patricklx commented on packages/vite/src/hbs.ts:
```diff
@@ -58,8 +59,8 @@ export function hbs(): Plugin {
-      let syntheticId = needsSyntheticComponentJS(source, resolution.id, resolverLoader.resolver.packageCache);
-      if (syntheticId) {
+      let syntheticId = needsSyntheticComponentJS(source, resolution.id);
+      if (syntheticId && isInComponents(resolution.id, resolverLoader.resolver.packageCache)) {
```
@ef4 is this okay like this?
